lidofinance/polygon-contracts
# fix: total matic to withdraw from the validators

# Summary
- Fix the `requestWithdraw` bug reported [here](https://github.com/lidofinance/polygon-contracts/blob/main/audits/v2/Shardlabs/LoP-v2-WithdrawRequest.pdf)
- Added recovery function.
- Added unit tests.
- simulation tests.

# Report
Here is the document that explains when the bug happened, its impact, and how to recover the protocol.
https://shardlabs.notion.site/RequestWithdraw-Report-0969548230254b1f8caf75b188bbf5bc

# Recovery
The report explains the impact of this bug on the protocol.
https://shardlabs.notion.site/RequestWithdraw-Report-0969548230254b1f8caf75b188bbf5bc

# Audit
- Audit the code fix related to requestWithdraw.
- Audit the code related to recovering function.

# Limitation
After adding the recovery function and the contract byte code (Max 24,576 bytes), to fix that I commented on 2 setter functions. 
- setPoLidoNFT: allows to set the NFT contract address.
- setVersion: allow to set the contract version.
The idea is to uncomment those 2 functions after deleting the recover function in a second upgrade. The PR is available here #15 
## Before
![before](https://user-images.githubusercontent.com/26064387/223672525-48b990b9-6c65-45fd-a4e0-af30a730d0e2.PNG)

## After
![after](https://user-images.githubusercontent.com/26064387/223672579-29c8b977-e42b-4e79-9a33-f49030aa6e38.PNG)

# Test
- Check if the requestWithdraw will cover the user request when the protocol is balanced and unbalanced [here](https://github.com/lidofinance/polygon-contracts/blob/6a9ae924cf86de0d18746530054a7ff8a0be30ab/test/NodeOperator.test.ts#L1294-L1366).
- Check if the rate exchange stMatic/Matic is correct when withdrawing from multiple validators [here](https://github.com/lidofinance/polygon-contracts/blob/6a9ae924cf86de0d18746530054a7ff8a0be30ab/test/StMatic.test.ts#L1329-L1375).
- Check the recovery logic function [here](https://github.com/lidofinance/polygon-contracts/blob/6a9ae924cf86de0d18746530054a7ff8a0be30ab/test/StMatic.test.ts#L2633-L2741).

# Simulation
- Fork mainnet and simulate the transactions executed after the upgrade to V2 (from `31-01-2023`) [here](https://github.com/lidofinance/polygon-contracts/blob/fix/requestWithdraw/simulation/simulateInFixedStMATIC.test.ts).
- Fork mainnet and Simulate the recovery [here](https://github.com/lidofinance/polygon-contracts/blob/fix/requestWithdraw/simulation/simulateRecover.test.ts).
